pageserver/src/tenant/storage_layer.rs

```diff
@@ -336,6 +311,952 @@ impl LayerAccessStats {
+                    let client = timeline
+                        .remote_client
+                        .as_ref()
+                        .expect("checked above with can_ever_evict");
+                    let result = client
+                        .download_layer_file(
+                            &this.desc.filename(),
+                            &crate::tenant::remote_timeline_client::index::LayerFileMetadata::new(
+                                this.desc.file_size,
+                            ),
+                        )
+                        .await;
+
+                    match result {
+                        Ok(size) => {
+                            timeline.metrics.resident_physical_size_gauge.add(size);
+                            let _ = tx.send(());
+                        }
+                        Err(e) => {
+                            // TODO: the temp file might still be around, metrics might be off
+                            tracing::error!("layer file download failed: {e:?}",);
+                        }
+                    }
+
+                    Ok(())
+                }
+                .in_current_span(),
+            );
+            if rx.await.is_err() {
+                return Err(anyhow::anyhow!("downloading failed, possibly for shutdown"));
+            }
+            // FIXME: we need backoff here so never spiral to download loop
+            anyhow::ensure!(
+                self.needs_download()
+                    .await
+                    .context("test if downloading is still needed")?
+                    .is_none(),
+                "post-condition for downloading: no longer needs downloading"
+            );
+        } else {
+            // the file is present locally and we could even be running without remote
+            // storage
+        }
+
+        // the assumption is that we own the layer residentness, no operator should go in
+        // and delete random files. this would be evident when trying to access the file
+        // Nth time (N>1) while having the VirtualFile evicted in between.
+        //
+        // we could support this by looping on NotFound from the layer access methods, but
+        // it's difficult to implement this so that the operator does not delete
+        // not-yet-uploaded files.
+
+        let res = Arc::new(DownloadedLayer {
+            owner: Arc::downgrade(self),
+            kind: tokio::sync::OnceCell::default(),
+        });
+
+        *locked = Some(if self.wanted_evicted.load(Ordering::Acquire) {
+            // because we reset wanted_evictness near beginning, this means when we were downloading someone
+            // wanted to evict this layer.
+            //
+            // perhaps the evict should only possible via ResidentLayer because this makes my head
+            // spin. the caller of this function will still get the proper `Arc<DownloadedLayer>`.
+            //
+            // the risk is that eviction becomes too flaky.
```

Comment:
Yes, but this leads to more eviction related tests to fail, because disk space is expectected to be freed but we are at the same time blocking the eviction.

Current impl establishes ResidentLayer only for the duration of "is this downloaded" check and then let's it go, and only `debug` tests fail. But that is not enough..

Reply:
This idea has now been bulldozed. `LayerE::evict_and_wait` should always do the right thing(tm), and I could actually write some unit tests for this as well.

Reply:
yes, it just occured to me that broadcast is not required, could just wait on a completion from the current *DownloadedLayer*.

Reply:
We cannot do that because the ... We could do it actually. It needs to be a completion "channel" and the sender is kept alive until the on_evict has completed... Problem is that it might still revert to Downloaded, except that it's not possible right now.

I guess I'll just leave it like this right now.